sql/sql_class.cc

```diff
@@ -1105,6 +1105,9 @@ THD::~THD() {
 extern "C" void thd_report_innodb_stat(THD *thd, unsigned long long trx_id,
                                        enum mysql_trx_stat_type type,
                                        unsigned long long value) {
+  if (!thd || thd_is_background_thread(thd))
```

Comment:
One thing I don't understand why catching a background thread or nullptr thd here works when it should have worked at an earlier point. In the scenario of this bug:
1. the crash comes from trx_stats::stop_lock_wait calling thd_report_innodb_stat with trx.mysql_thd == nullptr
2. But stop_lock_wait has an early return: if (UNIV_LIKELY(!take_stats)) return;
3. take_stats gets initialized at innobase_trx_init: trx->stats.set(innobase_slow_log_verbose(thd));
4. And innobase_slow_log_verbose already checks for nullptr and background thread thd: return thd && ... && !thd_is_background_thread(thd);

Why this does not work?

Reply:
This is/was also my question that I could not yet answer and is why I am unconvinced that this is the correct fix.  The thread 2/background thread seems to be not flagged as a background thread (thd_is_background_thread) and is determining that it should take_stats when the transaction is being created when it really should not.  I am still comparing to see where and why the difference between 5.7 as most of the rest of this appears to be unchanged in 8.0.

Reply:
Comparing with 5.7 wouldn't be my top choice how to proceed, chances are, the bug is the same here, but background thread difference makes it benign. I'd try dumping innodb_slow_log_verbose for thds in question at each step

Reply:
OK, I have a partial answer.  The trx used by thread 2 is not created through innobase_trx_init(...).  It is created directly from trx_allocate_for_background(...) -> trx_create_low(...) -> trx_pools->get(...) within dict0stats.cc::dict_stats_exec_sql(...) and is coming out with take_stats=true.

My theory now is that this may also be a bug in 5.7 and only through a specific sequence are we able to hit it in 8.0.  Looking at the way the pools work, there is no memory clearing going on in either get(...) or mem_free(...) processes the block (txn).  So I believe that in this very specific case, we are getting a previously used txn block out that had the take_stats=true and it is not being cleared at any point in the transaction 'allocation' and initialization process.  I did some brute force dumping of transaction, their real memory locations, some contents, etc... and I do see that the transaction that dict_stats_exec_sql got was just previously used by a client thread and it did in fact have take_stats=true.

Following  that thought process, I removed the current fix within thd_report_innodb_stat(...), re-verified the crash, then added  trx->stats.set(false); to trx_free(...) and this seems to have prevented the crash.

Will do more testing with this in place and possibly push it for more discussion.